nrf52-hal-common/src/spim.rs

```diff
@@ -158,61 +220,12 @@ where
 
     /// Internal helper function to setup and execute SPIM DMA transfer
     fn do_spi_dma_transfer(&mut self, tx: DmaSlice, rx: DmaSlice) -> Result<(), Error> {
-        // Conservative compiler fence to prevent optimizations that do not
-        // take in to account actions by DMA. The fence has been placed here,
-        // before any DMA action has started
-        compiler_fence(SeqCst);
+        let mut transfer = SpiSingleTransfer::new(&mut self.0, tx, rx);
 
-        // Set up the DMA write
-        self.0.txd.ptr.write(|w| unsafe { w.ptr().bits(tx.ptr) });
-
-        self.0.txd.maxcnt.write(|w|
-            // Note that that nrf52840 maxcnt is a wider
-            // type than a u8, so we use a `_` cast rather than a `u8` cast.
-            // The MAXCNT field is thus at least 8 bits wide and accepts the full
-            // range of values that fit in a `u8`.
-            unsafe { w.maxcnt().bits(tx.len as _ ) });
-
-        // Set up the DMA read
-        self.0.rxd.ptr.write(|w|
-            // This is safe for the same reasons that writing to TXD.PTR is
-            // safe. Please refer to the explanation there.
-            unsafe { w.ptr().bits(rx.ptr) });
-        self.0.rxd.maxcnt.write(|w|
-            // This is safe for the same reasons that writing to TXD.MAXCNT is
-            // safe. Please refer to the explanation there.
-            unsafe { w.maxcnt().bits(rx.len as _) });
-
-        // Start SPI transaction
-        self.0.tasks_start.write(|w|
-            // `1` is a valid value to write to task registers.
-            unsafe { w.bits(1) });
-
-        // Conservative compiler fence to prevent optimizations that do not
-        // take in to account actions by DMA. The fence has been placed here,
-        // after all possible DMA actions have completed
-        compiler_fence(SeqCst);
-
-        // Wait for END event
-        //
-        // This event is triggered once both transmitting and receiving are
-        // done.
-        while self.0.events_end.read().bits() == 0 {}
-
-        // Reset the event, otherwise it will always read `1` from now on.
-        self.0.events_end.write(|w| w);
-
-        // Conservative compiler fence to prevent optimizations that do not
-        // take in to account actions by DMA. The fence has been placed here,
-        // after all possible DMA actions have completed
-        compiler_fence(SeqCst);
-
-        if self.0.txd.amount.read().bits() != tx.len {
-            return Err(Error::Transmit);
-        }
-        if self.0.rxd.amount.read().bits() != rx.len {
-            return Err(Error::Receive);
+        while !transfer.poll_complete(&mut self.0)? {
+            core::sync::atomic::spin_loop_hint();
```

Comment:
This is not a spinlock, and I'm not sure this actually emits any instructions on thumbv7, so it probably can be removed (same for the others). If that's not the case and it should be kept, importing the function is a good idea.

Reply:
I added it as a hint to avoid https://github.com/rust-lang/rust/issues/28728.  I have seen that bug being triggered in embedded environments if the surrounding function (in this case `do_spi_dma_transfer`) is called from a function that diverges (i.e. has return type `!`), so I always add this as a precaution (but I'm not sure it always helps...).

In this case, it is a spinlock because all we are really doing is repeatedly doing a volatile read and waiting for a condition to be true (the holder of the lock just happens to be the hardware device, so to speak).  If it weren't a spinlock, we would have to be doing something in the loop instead of spinning (so any empty loop is sort of by definition a spinlock in my opinion, unless you do some IO logic in the actual `while` condition).

Another option here might be to do a `wfi` instead, do you think that would be sound?

I'll add imports in the meantime, at least.

Reply:
A volatile read is a sufficient side effect to avoid rust-lang/rust#28728. `wfi` would be sound, but it would simply lock up the processor if interrupts are disabled (which they might be as far as I can tell).

`spin_loop_hint` only really does anything if you have multiple hardware threads as far as I know. It might not even be sufficient for avoid that LLVM bug and only works by chance, like `loop { continue; }` seems to do today.

Reply:
Okay, not sure what happened in my case. I will remove the code and trust you!